src/platform/react-hooks/src/hooks/usePresenceEnter.ts

```diff
@@ -0,0 +1,59 @@
+import type * as Ably from 'ably';
+import { useCallback, useEffect } from 'react';
+import { ChannelParameters } from '../AblyReactHooks.js';
+import { useAbly } from './useAbly.js';
+import { useChannelInstance } from './useChannelInstance.js';
+import { useStateErrors } from './useStateErrors.js';
+
+export interface PresenceEnterResult<T> {
+  updateStatus: (messageOrPresenceObject: T) => void;
+  connectionError: Ably.ErrorInfo | null;
+  channelError: Ably.ErrorInfo | null;
+}
+
+const INACTIVE_CONNECTION_STATES: Ably.ConnectionState[] = ['suspended', 'closing', 'closed', 'failed'];
+
+export function usePresenceEnter<T = any>(
+  channelNameOrNameAndOptions: ChannelParameters,
+  messageOrPresenceObject?: T,
+): PresenceEnterResult<T> {
+  const params =
+    typeof channelNameOrNameAndOptions === 'object'
+      ? channelNameOrNameAndOptions
+      : { channelName: channelNameOrNameAndOptions };
+  const skip = params.skip;
+
+  const ably = useAbly(params.id);
+  const { channel } = useChannelInstance(params.id, params.channelName);
+  const { connectionError, channelError } = useStateErrors(params);
+
+  const onMount = async () => {
+    await channel.presence.enter(messageOrPresenceObject);
+  };
+
+  const onUnmount = () => {
+    // if connection is in one of inactive states, leave call will produce exception
+    if (channel.state === 'attached' && !INACTIVE_CONNECTION_STATES.includes(ably.connection.state)) {
+      channel.presence.leave();
```

Comment:
I know this was the behaviour in the previous `usePresence` hook, but i'm not sure if this behaviour is more helpful than not - for example if you have two components mounted with this hook and one of them is umounted then you'll leave presence even though you still have a component with `usePresenceEnter` mounted. I do think it would be easier to support users explicitly entering/leaving presence, but if we are gonna provide a hook it should have a more intelligent way of automatically leaving presence than just every component unmount. WDYT @VeskeR?

Reply:
I thought a bit about this earlier today too.
We may need something like it's done in `useChannel` with its channel detachment logic:
https://github.com/ably/ably-js/blob/46dcda7f69350657ef5579344dfe95bd2ca0b967/src/platform/react-hooks/src/hooks/useChannel.ts#L92
There we check the count of existing listeners on the channel, so if there are multiple `useChannel` hooks and some of them unmount, we won't detach the channel, since there are still listeners active.

So one option would be to keep track of mounted `usePresenceEnter` hooks and check that we unmounted every single one of them before leaving the presence.

> I do think it would be easier to support users explicitly entering/leaving presence

This is something I didn't see discussed before. It would be kinda similar to how we added 'publish' function to `useChannel` hook in https://github.com/ably/ably-js/pull/1658, we would have something like:
```tsx
const { enter, leave } = usePresence('channel-name');

useEffect(() => {
  enter();

  return () => {
    leave();
  };
}, [enter, leave]);
```
so users can have more granular control at which point in their app they want to enter presence, and when exactly they would like to leave it.

Reply:
Just for the record, `useChannel` detach logic is also flawed, e.g. I believe `channel.listeners.length` will be 0 if you only subscribe to presence updates, but channel shouldn't be detached. And I think now we should move this logic to the `ChannelProvider`.

Explicitly entering or leaving presence may work; the only thing that concerns me is that this way, we have less control, and the user should take care of this themselves, which could potentially lead to a wrong implementation.

I started feeling maybe we need `PresenceProvider` as well :) The problem with current `usePresence` hooks that even if we calculate all refs and leave only when it's necessary, we still have no control when they are rendered and if they have different payload for `enter` will have different results.

Reply:
But on second thoughts I don't know how to design `PresenceProvider`, I don't want it to enter presence immediately, this way we have to implement `skip` parameter and it will look weird

Reply:
I think I'll vote for some ref counter for entering/leaving

Reply:
Most likely will implement ref counter to handle leaving presence intelligently. Created https://github.com/ably/ably-js/issues/1687 for this, won't be a part of this PR.